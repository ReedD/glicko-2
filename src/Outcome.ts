enum Outcome {
  Win = 1,
  Loss = 0,
  Tie = 0.5,
}

export const negateOutcome = (outcome: Outcome) => {
  switch (outcome) {
    case Outcome.Win:
      return Outcome.Loss;
    case Outcome.Loss:
      return Outcome.Win;
    default:
      return Outcome.Tie;
  }
};

export default Outcome;
