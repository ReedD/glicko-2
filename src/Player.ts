import Outcome from './Outcome';

export interface IPlayerOpts {
  defaultRating: number;
  rating: number;
  ratingDeviation: number;
  tau: number;
  volatility: number;
}

export default class Player {
  private static readonly scalingFactor = 173.7178;

  private readonly tau: number;
  private readonly defaultRating: number;

  private _rating: number;
  private _ratingDeviation: number;
  private _volatility: number;
  private opponentRatings: number[] = [];
  private opponentRatingDeviations: number[] = [];
  private outcomes: Outcome[] = [];

  constructor({
    defaultRating,
    rating,
    ratingDeviation,
    tau,
    volatility,
  }: IPlayerOpts) {
    this.tau = tau;
    this.defaultRating = defaultRating;
    this._rating = (rating - defaultRating) / Player.scalingFactor;
    this._ratingDeviation = ratingDeviation / Player.scalingFactor;
    this._volatility = volatility;
  }

  static compositePlayer(players: Player[]) {
    if (players.length === 0) {
      throw new Error('Cannot create a composite of 0 players');
    }
    const [refPlayer] = players;
    let rating = 0;
    let ratingDeviation = 0;
    players.forEach((p, i) => {
      if (!refPlayer.isCompatiblePlayer(p)) {
        throw new Error('All players must have equal tau and defaultRating');
      }
      // Cumulative moving average
      rating = rating + (p.rating - rating) / (i + 1);
      ratingDeviation =
        ratingDeviation + (p.ratingDeviation - ratingDeviation) / (i + 1);
    });
    return new Player({
      defaultRating: refPlayer.defaultRating,
      rating,
      ratingDeviation,
      tau: refPlayer.tau,
      // This is a pseudo player whose resulting volatility is irrelevant
      volatility: refPlayer.volatility,
    });
  }

  get rating() {
    return this._rating * Player.scalingFactor + this.defaultRating;
  }

  get ratingDeviation() {
    return this._ratingDeviation * Player.scalingFactor;
  }

  get volatility() {
    return this._volatility;
  }

  addResult(opponent: Player, outcome: Outcome) {
    this.opponentRatings.push(opponent._rating);
    this.opponentRatingDeviations.push(opponent._ratingDeviation);
    this.outcomes.push(outcome);
  }

  // Calculates the new rating and rating deviation of the player.
  // Follows the steps of the algorithm described here:
  // http://www.glicko.net/glicko/glicko2.pdf
  updateRating() {
    if (!this.hasPlayed()) {
      // Applies only the Step 6 of the algorithm
      this.preRatingDeviation();
      return;
    }

    // Step 1: done by Player initialization
    // Step 2: done by set _rating and set _ratingDeviation

    // Step 3
    const v = this.variance();

    // Step 4
    const delta = this.delta(v);

    // Step 5
    this._volatility = this.volatilityAlgorithm(v, delta);

    // Step 6
    this.preRatingDeviation();

    // Step 7
    this._ratingDeviation =
      1 / Math.sqrt(1 / Math.pow(this._ratingDeviation, 2) + 1 / v);

    let tempSum = 0;
    for (let i = 0, len = this.opponentRatings.length; i < len; i++) {
      tempSum +=
        this.g(this.opponentRatingDeviations[i]) *
        (this.outcomes[i] -
          this.E(this.opponentRatings[i], this.opponentRatingDeviations[i]));
    }
    this._rating += Math.pow(this._ratingDeviation, 2) * tempSum;

    // Step 8: done by getters of `rating` and `ratingDeviation`

    // Cleanup
    this.cleanPreviousMatches();
  }

  private isCompatiblePlayer(player: Player) {
    return (
      player.tau === this.tau && player.defaultRating === this.defaultRating
    );
  }

  private cleanPreviousMatches() {
    this.opponentRatings = [];
    this.opponentRatingDeviations = [];
    this.outcomes = [];
  }

  private hasPlayed() {
    return this.outcomes.length > 0;
  }

  private volatilityAlgorithm(v: number, delta: number) {
    // Step 5.1
    let A = Math.log(Math.pow(this._volatility, 2));
    const f = this.fFactory(delta, v, A);
    const epsilon = 0.000001;

    // Step 5.2
    let B: number, k: number;
    if (Math.pow(delta, 2) > Math.pow(this._ratingDeviation, 2) + v) {
      B = Math.log(Math.pow(delta, 2) - Math.pow(this._ratingDeviation, 2) - v);
    } else {
      k = 1;
      while (f(A - k * this.tau) < 0) {
        k = k + 1;
      }
      B = A - k * this.tau;
    }

    // Step 5.3
    let fA = f(A);
    let fB = f(B);

    // Step 5.4
    let C: number, fC: number;
    while (Math.abs(B - A) > epsilon) {
      C = A + ((A - B) * fA) / (fB - fA);
      fC = f(C);
      if (fC * fB < 0) {
        A = B;
        fA = fB;
      } else {
        fA = fA / 2;
      }
      B = C;
      fB = fC;
    }
    // Step 5.5
    return Math.exp(A / 2);
  }

  // Calculates and updates the player's rating deviation for the beginning of a
  // rating period.
  private preRatingDeviation() {
    this._ratingDeviation = Math.sqrt(
      Math.pow(this._ratingDeviation, 2) + Math.pow(this._volatility, 2),
    );
  }

  // Calculation of the estimated variance of the player's rating based on game
  // outcomes
  private variance() {
    let tempSum = 0;
    for (let i = 0, len = this.opponentRatings.length; i < len; i++) {
      const tempE = this.E(
        this.opponentRatings[i],
        this.opponentRatingDeviations[i],
      );
      tempSum +=
        Math.pow(this.g(this.opponentRatingDeviations[i]), 2) *
        tempE *
        (1 - tempE);
    }
    return 1 / tempSum;
  }

  // The Glicko E function.
  private E(p2rating: number, p2ratingDeviation: number) {
    return (
      1 /
      (1 + Math.exp(-1 * this.g(p2ratingDeviation) * (this._rating - p2rating)))
    );
  }

  // The Glicko2 g(ratingDeviation) function.
  private g(ratingDeviation: number) {
    return (
      1 /
      Math.sqrt(1 + (3 * Math.pow(ratingDeviation, 2)) / Math.pow(Math.PI, 2))
    );
  }

  // The delta function of the Glicko2 system.
  // Calculation of the estimated improvement in rating (step 4 of the algorithm)
  private delta(v: number) {
    let tempSum = 0;
    for (let i = 0, len = this.opponentRatings.length; i < len; i++) {
      tempSum +=
        this.g(this.opponentRatingDeviations[i]) *
        (this.outcomes[i] -
          this.E(this.opponentRatings[i], this.opponentRatingDeviations[i]));
    }
    return v * tempSum;
  }

  private fFactory(delta: number, v: number, a: number) {
    return (x: number) => {
      return (
        (Math.exp(x) *
          (Math.pow(delta, 2) -
            Math.pow(this._ratingDeviation, 2) -
            v -
            Math.exp(x))) /
          (2 *
            Math.pow(Math.pow(this._ratingDeviation, 2) + v + Math.exp(x), 2)) -
        (x - a) / Math.pow(this.tau, 2)
      );
    };
  }

  toObject() {
    return {
      rating: this.rating,
      ratingDeviation: this.ratingDeviation,
      volatility: this.volatility,
    };
  }

  toJSON() {
    return this.toObject();
  }
}

export const createPlayerFactory = ({
  defaultRating = 1500,
  defaultRatingDeviation = 350,
  defaultVolatility = 0.06,
  tau = 0.5,
} = {}) => ({
  rating = defaultRating,
  ratingDeviation = defaultRatingDeviation,
  volatility = defaultVolatility,
} = {}) =>
  new Player({
    defaultRating,
    rating,
    ratingDeviation,
    tau,
    volatility,
  });
