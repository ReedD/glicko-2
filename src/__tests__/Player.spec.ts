import Outcome from '../Outcome';
import Player, { createPlayerFactory } from '../Player';
import { serializePlayer } from './utils';

describe('Player', () => {
  const createPlayer = createPlayerFactory();

  describe('createPlayerFactory', () => {
    it('should create new playerFactory', () => {
      const createCustomPlayer = createPlayerFactory({
        defaultRating: 1600,
        defaultRatingDeviation: 200,
        defaultVolatility: 0.09,
        tau: 0.8,
      });
      const player = createCustomPlayer();
      expect(serializePlayer(player)).toMatchSnapshot();
    });

    it('should not be affected by other Player factories', () => {
      const createCustomPlayer = createPlayerFactory({ defaultRating: 1600 });
      const customPlayer = createCustomPlayer();
      const player = createPlayer();
      expect(player.rating).toBe(1500);
      expect(customPlayer.rating).toBe(1600);
    });
  });

  describe('compositePlayer', () => {
    it('should make a default player when passed no settings', () => {
      const aPlayer = createPlayer();
      const bPlayer = createPlayer({ rating: 1600, ratingDeviation: 300 });
      const abPlayer = Player.compositePlayer([aPlayer, bPlayer]);
      expect(abPlayer).toMatchSnapshot();
    });

    it('should fail to make a composite of with empty array', () => {
      expect(() => {
        Player.compositePlayer([]);
      }).toThrowErrorMatchingSnapshot();
    });

    it('should fail to make a composite of two incompatible players', () => {
      const createCustomPlayer = createPlayerFactory({ defaultRating: 1600 });
      const aPlayer = createPlayer();
      const bPlayer = createCustomPlayer();
      expect(() => {
        Player.compositePlayer([aPlayer, bPlayer]);
      }).toThrowErrorMatchingSnapshot();
    });
  });

  describe('constructor', () => {
    it('should make a default player when passed no settings', () => {
      const player = createPlayer();
      expect(player).toMatchSnapshot();
    });

    it('should support setting individual settings', () => {
      const player = createPlayer({ rating: 1600 });
      expect(player).toMatchSnapshot();
    });

    it('should throw if invalid player', () => {
      expect(() =>
        createPlayer({ outcomes: [1] }),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        createPlayer({ opponentRatingDeviations: [1] }),
      ).toThrowErrorMatchingSnapshot();
      expect(() =>
        createPlayer({ opponentRatings: [1] }),
      ).toThrowErrorMatchingSnapshot();
    });
  });

  describe('addResult', () => {
    it('should add a result with no update to rating or opponent', () => {
      const ryan = createPlayer();
      const bob = createPlayer();
      ryan.addResult(bob, Outcome.Win);
      ryan.addResult(bob, Outcome.Loss);
      ryan.addResult(bob, Outcome.Tie);
      expect(serializePlayer(ryan)).toMatchSnapshot();
      expect(serializePlayer(bob)).toMatchSnapshot();
      expect(ryan).toMatchSnapshot();
      expect(bob).toMatchSnapshot();
    });
  });

  describe('updateRatings', () => {
    it('should calculate new ratings', () => {
      // Following the example calculation given here:
      // http://www.glicko.net/glicko/glicko2.pdf
      // Suppose a player rated 1500 competes against players rated 1400, 1550
      // and 1700, winning the first game and losing the next two. Assume the
      // 1500-rated player's rating deviation is 200, and his opponents' are 30,
      // 100 and 300, respectively. Assume the 1500 player has volatility
      // σ = 0.06, and the system constant τ is 0.5.

      const ryan = createPlayer({ ratingDeviation: 200 });
      const bob = createPlayer({
        rating: 1400,
        ratingDeviation: 30,
        volatility: 0.06,
      });
      const john = createPlayer({
        rating: 1550,
        ratingDeviation: 100,
        volatility: 0.06,
      });
      const mary = createPlayer({
        rating: 1700,
        ratingDeviation: 300,
        volatility: 0.06,
      });

      ryan.addResult(bob, Outcome.Win);
      ryan.addResult(john, Outcome.Loss);
      ryan.addResult(mary, Outcome.Loss);

      ryan.updateRating();

      expect(ryan.rating).toBeCloseTo(1464.06, 1);
      expect(ryan.ratingDeviation).toBeCloseTo(151.52, 2);
      expect(ryan.volatility).toBeCloseTo(0.05999, 4);
    });

    it('should be allowed to be called multiple times', () => {
      const ryan = createPlayer({ ratingDeviation: 200 });
      const bob = createPlayer({
        rating: 1400,
        ratingDeviation: 30,
        volatility: 0.06,
      });
      const john = createPlayer({
        rating: 1550,
        ratingDeviation: 100,
        volatility: 0.06,
      });
      const mary = createPlayer({
        rating: 1700,
        ratingDeviation: 300,
        volatility: 0.06,
      });

      ryan.addResult(bob, Outcome.Win);
      ryan.addResult(john, Outcome.Loss);
      ryan.addResult(mary, Outcome.Loss);

      ryan.updateRating();
      bob.updateRating();
      john.updateRating();
      mary.updateRating();

      // Initiate clones of each player
      const ryan1 = createPlayer(ryan);
      const bob1 = createPlayer(bob);
      const john1 = createPlayer(john);
      const mary1 = createPlayer(mary);

      // Second tournament for the original players
      ryan.addResult(bob, Outcome.Loss);
      ryan.addResult(john, Outcome.Win);
      mary.addResult(bob, Outcome.Win);

      ryan.updateRating();
      bob.updateRating();
      john.updateRating();
      mary.updateRating();

      // Replay second tournament results against the clones
      ryan1.addResult(bob1, Outcome.Loss);
      ryan1.addResult(john1, Outcome.Win);
      mary1.addResult(bob1, Outcome.Win);

      ryan1.updateRating();
      bob1.updateRating();
      john1.updateRating();
      mary1.updateRating();

      // The ratings in both systems should be the same
      expect(ryan.rating).toBe(ryan1.rating);
      expect(ryan.ratingDeviation).toBe(ryan1.ratingDeviation);
      expect(ryan.volatility).toBe(ryan1.volatility);
    });

    it('should be able to update ratings when a player did not play', () => {
      const ryan = createPlayer();
      ryan.updateRating();
      expect(ryan.rating).toBeCloseTo(1500);
      expect(ryan.ratingDeviation).toBeCloseTo(350.155);
      expect(ryan.volatility).toBeCloseTo(0.06);
    });

    it('should clear internals after update', () => {
      const ryan = createPlayer();
      const bob = createPlayer();
      ryan.addResult(bob, Outcome.Win);
      ryan.addResult(bob, Outcome.Loss);
      ryan.addResult(bob, Outcome.Tie);
      expect(serializePlayer(ryan)).toMatchSnapshot();
      ryan.updateRating();
      expect(serializePlayer(ryan)).toMatchSnapshot();
    });
  });
});
