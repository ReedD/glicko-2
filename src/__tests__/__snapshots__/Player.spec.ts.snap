// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player addResult should add a result with no update to rating or opponent 1`] = `
Object {
  "_rating": 0,
  "_ratingDeviation": 2.01,
  "_volatility": 0.06,
  "defaultRating": 1500,
  "opponentRatingDeviations": Array [
    2.01,
    2.01,
    2.01,
  ],
  "opponentRatings": Array [
    0,
    0,
    0,
  ],
  "outcomes": Array [
    1,
    0,
    0.5,
  ],
  "rating": 1500,
  "ratingDeviation": 350,
  "tau": 0.5,
  "volatility": 0.06,
}
`;

exports[`Player addResult should add a result with no update to rating or opponent 2`] = `
Object {
  "_rating": 0,
  "_ratingDeviation": 2.01,
  "_volatility": 0.06,
  "defaultRating": 1500,
  "opponentRatingDeviations": Array [],
  "opponentRatings": Array [],
  "outcomes": Array [],
  "rating": 1500,
  "ratingDeviation": 350,
  "tau": 0.5,
  "volatility": 0.06,
}
`;

exports[`Player addResult should add a result with no update to rating or opponent 3`] = `
Object {
  "rating": 1500,
  "ratingDeviation": 350,
  "volatility": 0.06,
}
`;

exports[`Player addResult should add a result with no update to rating or opponent 4`] = `
Object {
  "rating": 1500,
  "ratingDeviation": 350,
  "volatility": 0.06,
}
`;

exports[`Player compositePlayer should fail to make a composite of two incompatible players 1`] = `"All players must have equal tau and defaultRating"`;

exports[`Player compositePlayer should fail to make a composite of with empty array 1`] = `"Cannot create a composite of 0 players"`;

exports[`Player compositePlayer should make a default player when passed no settings 1`] = `
Object {
  "rating": 1550,
  "ratingDeviation": 325,
  "volatility": 0.06,
}
`;

exports[`Player constructor should make a default player when passed no settings 1`] = `
Object {
  "rating": 1500,
  "ratingDeviation": 350,
  "volatility": 0.06,
}
`;

exports[`Player constructor should support setting individual settings 1`] = `
Object {
  "rating": 1600,
  "ratingDeviation": 350,
  "volatility": 0.06,
}
`;

exports[`Player createPlayerFactory should create new playerFactory 1`] = `
Object {
  "_rating": 0,
  "_ratingDeviation": 1.15,
  "_volatility": 0.09,
  "defaultRating": 1600,
  "opponentRatingDeviations": Array [],
  "opponentRatings": Array [],
  "outcomes": Array [],
  "rating": 1600,
  "ratingDeviation": 200,
  "tau": 0.8,
  "volatility": 0.09,
}
`;

exports[`Player updateRatings should clear internals after update 1`] = `
Object {
  "_rating": 0,
  "_ratingDeviation": 2.01,
  "_volatility": 0.06,
  "defaultRating": 1500,
  "opponentRatingDeviations": Array [
    2.01,
    2.01,
    2.01,
  ],
  "opponentRatings": Array [
    0,
    0,
    0,
  ],
  "outcomes": Array [
    1,
    0,
    0.5,
  ],
  "rating": 1500,
  "ratingDeviation": 350,
  "tau": 0.5,
  "volatility": 0.06,
}
`;

exports[`Player updateRatings should clear internals after update 2`] = `
Object {
  "_rating": 0,
  "_ratingDeviation": 1.31,
  "_volatility": 0.06,
  "defaultRating": 1500,
  "opponentRatingDeviations": Array [],
  "opponentRatings": Array [],
  "outcomes": Array [],
  "rating": 1500,
  "ratingDeviation": 227.74,
  "tau": 0.5,
  "volatility": 0.06,
}
`;
